File: app/src/main/java/knf/kuma/favorite/objects/InfoContainer.java
package knf.kuma.favorite.objects;

import java.util.ArrayList;
import java.util.List;

import knf.kuma.pojos.FavSection;
import knf.kuma.pojos.FavoriteObject;

public class InfoContainer {
    public List<FavoriteObject> updated;
    public boolean needReload = false;
    public int from;
    public int to;
    private List<FavoriteObject> current;

    public InfoContainer() {

    }

    public void setLists(List<FavoriteObject> current, List<FavoriteObject> updated) {
        this.current = new ArrayList<>(current);
        this.updated = new ArrayList<>(updated);
    }

    public void reload(FavoriteObject object) {
        if (object instanceof FavSection || object == null) {
            needReload = true;
        } else if (!updated.contains(object)) {
            needReload = true;
        } else if (current.size() != updated.size()) {
            needReload = true;
        } else {
            needReload = false;
            from = current.indexOf(object);
            to = updated.indexOf(object);
        }
    }
}

File: app/src/main/java/knf/kuma/favorite/objects/InfoContainer.kt
package knf.kuma.favorite.objects

import knf.kuma.pojos.FavSection
import knf.kuma.pojos.FavoriteObject
import java.util.*

class InfoContainer {
    var updated: MutableList<FavoriteObject>? = arrayListOf()
    var needReload = false
    var from: Int = 0
    var to: Int = 0
    private var current: MutableList<FavoriteObject>? = null

    fun setLists(current: MutableList<FavoriteObject>, updated: MutableList<FavoriteObject>) {
        this.current = ArrayList(current)
        this.updated = ArrayList(updated)
    }

    fun reload(favoriteObject: FavoriteObject?) {
        when {
            favoriteObject == null -> needReload = true
            favoriteObject is FavSection -> needReload = true
            updated?.contains(favoriteObject) == false -> needReload = true
            current?.size != updated?.size -> needReload = true
            else -> {
                needReload = false
                from = current?.indexOf(favoriteObject) ?: -1
                to = updated?.indexOf(favoriteObject) ?: -1
            }
        }
    }
}
